File: application/src/test/java/com/catalog/application/category/create/CreateCategoryUseCaseTest.java
package com.catalog.application.category.create;

import com.catalog.domain.category.CategoryGateway;
import com.catalog.domain.exceptions.DomainException;
import com.catalog.domain.validation.handler.Notification;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Objects;

import static org.mockito.AdditionalAnswers.returnsFirstArg;

@ExtendWith(MockitoExtension.class)
public class CreateCategoryUseCaseTest {
    @InjectMocks
    private DefaultCreateCategoryUseCase useCase;

    @Mock
    private CategoryGateway gateway;

    @BeforeEach
    void cleanUp() {
        Mockito.reset(this.gateway);
    }

    @Test
    public void givenAValidCommand_whenCallsCreateCategory_shouldReturnCategoryId() {
        final var expectName = "Movies";
        final var expectDescription = "Category of movies";
        final var expectIsActive = true;

        final var aCommand = CreateCategoryCommand.with(expectName, expectDescription, expectIsActive);

        Mockito.when(gateway.create(Mockito.any())).thenAnswer(returnsFirstArg());

        final var actualOutput = useCase.execute(aCommand).get();

        Assertions.assertNotNull(actualOutput);
        Assertions.assertNotNull(actualOutput.id());

        Mockito.verify(gateway, Mockito.times(1))
                .create(Mockito.argThat(aCategory ->
                                Objects.equals(expectName, aCategory.getName())
                                        && Objects.equals(expectDescription, aCategory.getDescription())
                                        && Objects.equals(expectIsActive, aCategory.isActive()
                                        && Objects.nonNull(aCategory.getId())
                                        && Objects.nonNull(aCategory.getCreatedAt())
                                        && Objects.nonNull(aCategory.getUpdatedAt())
                                        && Objects.isNull(aCategory.getDeletedAt())
                                )
                        ));
    }

    @Test
    public void givenAInvalidName_whenCallsCreateCategory_shouldReturnDomainException() {
        final var expectName = " ";
        final var expectDescription = "Category of movies";
        final var expectIsActive = true;

        final var aCommand = CreateCategoryCommand.with(expectName, expectDescription, expectIsActive);

        final var notification = useCase.execute(aCommand).getLeft();

        Assertions.assertEquals(1, notification.getErrors().size());
        Assertions.assertEquals("'name' should not be empty", notification.firstError().message());

        Mockito.verify(gateway, Mockito.times(0)).create(Mockito.any());
    }

    @Test
    public void givenAValidCommandWithInactiveCategory_whenCallsCreateCategory_shouldReturnCategoryId() {
        final var expectName = "Movies";
        final var expectDescription = "Category of movies";
        final var expectIsActive = false;

        final var aCommand = CreateCategoryCommand.with(expectName, expectDescription, expectIsActive);

        Mockito.when(gateway.create(Mockito.any())).thenAnswer(returnsFirstArg());

        final var actualOutput = useCase.execute(aCommand).get();

        Assertions.assertNotNull(actualOutput);
        Assertions.assertNotNull(actualOutput.id());

        Mockito.verify(gateway, Mockito.times(1))
                .create(Mockito.argThat(aCategory ->
                                Objects.equals(expectName, aCategory.getName())
                                        && Objects.equals(expectDescription, aCategory.getDescription())
                                        && Objects.equals(expectIsActive, aCategory.isActive())
                                        && Objects.nonNull(aCategory.getId())
                                        && Objects.nonNull(aCategory.getCreatedAt())
                                        && Objects.nonNull(aCategory.getUpdatedAt())
                                        && Objects.nonNull(aCategory.getDeletedAt())
                        ));
    }

    @Test
    public void givenAValidCommand_whenGatewayThrowsRandomException_shouldReturnAnException() {
        final var expectName = "Movies";
        final var expectDescription = "Category of movies";
        final var expectIsActive = false;

        final var aCommand = CreateCategoryCommand.with(expectName, expectDescription, expectIsActive);

        Mockito.when(gateway.create(Mockito.any())).thenThrow(new IllegalStateException("Gateway error"));

        final var notification = useCase.execute(aCommand).getLeft();

        Assertions.assertEquals(1, notification.getErrors().size());
        Assertions.assertEquals("Gateway error", notification.firstError().message());

        Mockito.verify(gateway, Mockito.times(1))
                .create(Mockito.argThat(aCategory ->
                        Objects.equals(expectName, aCategory.getName())
                                && Objects.equals(expectDescription, aCategory.getDescription())
                                && Objects.equals(expectIsActive, aCategory.isActive())
                                && Objects.nonNull(aCategory.getId())
                                && Objects.nonNull(aCategory.getCreatedAt())
                                && Objects.nonNull(aCategory.getUpdatedAt())
                                && Objects.nonNull(aCategory.getDeletedAt())
                ));
    }
}

File: application/src/test/kotlin/com/full/admin/catalog/application/category/create/CreateCategoryUseCaseTest.kt
package com.full.admin.catalog.application.category.create

import arrow.core.*
import category.Category
import category.CategoryGateway
import exceptions.DomainException
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.mockk
import io.mockk.verify
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import kotlin.test.BeforeTest

@ExtendWith(MockKExtension::class)
class CreateCategoryUseCaseTest {

    @InjectMockKs
    private lateinit var useCase: DefaultCreateCategoryUseCase

    @MockK
    private lateinit var categoryGateway: CategoryGateway

    @BeforeTest
    fun setUp() = MockKAnnotations.init(this, relaxUnitFun = true)

    @Test
    fun givenAValidCommand_whenCallsCreateCategory_shouldReturnCategoryId() {
        val expectedName = "Movies"
        val expectedDescription = "The category most watched"
        val expectedIsActive = true

        val aCommand = CreateCategoryCommand.with(expectedName, expectedDescription, expectedIsActive)

        every { categoryGateway.create(any()) } answers { firstArg() }

        val actualOutput = useCase.execute(aCommand)

        Assertions.assertTrue(actualOutput.isRight())

        val rightValue = actualOutput.getOrElse { throw AssertionError("Expected Right, got Left") }
        Assertions.assertNotNull(rightValue.id)

        verify(exactly = 1) { categoryGateway.create(match { category: Category ->
            category.name == expectedName
                    && category.description == expectedDescription
                    && category.isActive == expectedIsActive
                    && category.deletedAt == null
        }) }
    }

    @Test
    fun givenAnInvalidName_thenCallCreateCategory_shouldReturnDomainException() {
        val expectedName = "    "
        val expectedDescription = "The category most watched"
        val expectedIsActive = true
        val expectedErrorMessage = "'name' should not be empty"
        val expectedErrorCount = 1

        val aCommand = CreateCategoryCommand.with(expectedName, expectedDescription, expectedIsActive)

        val actualOutput = useCase.execute(aCommand)

        Assertions.assertTrue(actualOutput.isLeft())

        val actualException = actualOutput.swap().getOrElse { throw AssertionError("Expected Right, got Left") }

        Assertions.assertEquals(expectedErrorCount, actualException.getErrors().size)
        Assertions.assertEquals(expectedErrorMessage, actualException.firstError()?.message)

        verify(exactly = 0) { categoryGateway.create(any()) }
    }

    @Test
    fun givenAValidCommandWithInactiveCategory_thenCallCreateCategory_shouldReturnInactiveCategoryId() {
        val expectedName = "Movies"
        val expectedDescription = "The category most watched"
        val expectedIsActive = false

        val aCommand = CreateCategoryCommand.with(expectedName, expectedDescription, expectedIsActive)

        every { categoryGateway.create(any()) } answers { firstArg() }

        val actualOutput = useCase.execute(aCommand)

        val rightValue = actualOutput.getOrElse { throw AssertionError("Expected Right, got Left") }
        Assertions.assertNotNull(rightValue.id)

        verify(exactly = 1) { categoryGateway.create(match { category: Category ->
            category.name == expectedName
                    && category.description == expectedDescription
                    && category.isActive == expectedIsActive
                    && category.deletedAt != null
        }) }
    }

    @Test
    fun givenAValidCommand_whenGatewayThrowsRandomException_shouldReturnAnException() {
        val expectedName = "Movies"
        val expectedDescription = "The category most watched"
        val expectedIsActive = true
        val expectedErrorMessage = "Gateway error"
        val expectedErrorCount = 1

        val aCommand = CreateCategoryCommand.with(expectedName, expectedDescription, expectedIsActive)

        every { categoryGateway.create(any()) } throws IllegalStateException(expectedErrorMessage)

        val actualOutput = useCase.execute(aCommand)

        Assertions.assertTrue(actualOutput.isLeft())

        val actualException = actualOutput.swap().getOrElse { throw AssertionError("Expected Right, got Left") }

        Assertions.assertEquals(expectedErrorCount, actualException.getErrors().size)
        Assertions.assertEquals(expectedErrorMessage, actualException.firstError()?.message)

        verify(exactly = 1) { categoryGateway.create(match { category: Category ->
            category.name == expectedName
                    && category.description == expectedDescription
                    && category.isActive == expectedIsActive
                    && category.deletedAt == null
        }) }
    }
}